src/main.rs

```diff
@@ -1,10 +1,64 @@
 use std::env;
+use std::io::{self, Write};
+use std::process;
+use std::borrow::Cow;
 
-fn main() {
-    let arguments = env::args().collect::<Vec<_>>();
-    
-    match arguments.len() {
-        1 => loop { println!("y") },
-        _ => loop { println!("{}", arguments[1]) },
+#[cfg(not(unix))]
+mod platform {
+    use std::ffi::OsString;
+    pub const BUFFER_CAPACITY: usize = 16 * 1024;
+
+    pub fn to_bytes(os_str: OsString) -> Vec<u8> {
+        os_str.into_string().expect("non utf-8 argument only supported on unix").into()
+    }
+}
+
+#[cfg(unix)]
+mod platform {
+    use std::ffi::OsString;
+    pub const BUFFER_CAPACITY: usize = 64 * 1024;
+
+    pub fn to_bytes(os_str: OsString) -> Vec<u8> {
+        use std::os::unix::ffi::OsStringExt;
+        os_str.into_vec()
     }
 }
+
+use platform::*;
+
+fn fill_up_buffer<'a>(buffer: &'a mut [u8], output: &'a [u8]) -> &'a [u8] {
+    if output.len() > buffer.len() / 2 {
+        return output;
+    }
+
+    let mut buffer_size = output.len();
+    buffer[..buffer_size].clone_from_slice(output);
+
+    while buffer_size < buffer.len() {
+        let (left, mut right) = buffer.split_at_mut(buffer_size);
+        right.write(left).unwrap();
+        buffer_size *= 2;
+    }
+
+    &buffer[..buffer_size]
+}
+
+
+fn write(output: &[u8]) {
+    let stdout = io::stdout();
+    let mut locked = stdout.lock();
+    let mut buffer = [0u8; BUFFER_CAPACITY];
+
+    while locked.write_all(fill_up_buffer(&mut buffer, output)).is_ok() {}
```

Comment:
If LLVM already does optimize it, I'm not sure there's much reason to do so manually. I might look at it this evening, but I don't have access to a rustc right now.